File: src/main/java/ch20_generics/DogsAndRobots.py
# generics/DogsAndRobots.py
# (c)2021 MindView LLC: see Copyright.txt
# We make no guarantees that this code is fit for any purpose.
# Visit http://OnJava8.com for more book information.
# P.731 §20.15 潜在类型机制 §20.15.1 Python 中的潜在类型机制

class Dog:
    def speak(self):
        print("Arf!")
    def sit(self):
        print("Sitting")
    def reproduce(self):
        pass

class Robot:
    def speak(self):
        print("Click!")
    def sit(self):
        print("Clank!")
    def oilChange(self):
        pass
"""
【Python：动态潜在类型机制的极致灵活】
本例展示了最纯粹的“鸭子类型（Duck Typing）”：只关注对象是否具备特定方法（speak/sit），完全忽略对象的继承关系与类型标记。
这与 Java 泛型强制要求的“名义类型（Nominal Typing）”形成鲜明对比——Java 必须通过显式的接口继承（extends Interface）来赋予类型资格。
Python 证明了代码复用的最高境界是不依赖任何特定层级结构，虽将类型检查推迟至运行时，却最大限度地暴露了 Java 泛型因过度追求编译期边界而丧失的灵活性。
"""
def perform(anything):
    anything.speak()
    anything.sit()

a = Dog()
b = Robot()
perform(a)
perform(b)

output = """
Arf!
Sitting
Click!
Clank!
"""

File: src/main/java/ch20_generics/dogsandrobots.go
// generics/dogsandrobots.go
// (c)2021 MindView LLC: see Copyright.txt
// We make no guarantees that this code is fit for any purpose.
// Visit http://OnJava8.com for more book information.
// P.734 §20.15 潜在类型机制 §20.15.3 Go 中的潜在类型机制

package main
import "fmt"

type Dog struct {}
func (this Dog) speak() { fmt.Printf("Arf!\n")}
func (this Dog) sit() { fmt.Printf("Sitting\n")}
func (this Dog) reproduce() {}

type Robot struct {}
func (this Robot) speak() { fmt.Printf("Click!\n") }
func (this Robot) sit() { fmt.Printf("Clank!\n") }
func (this Robot) oilChange() {}
/*
 * 【Go：结构化类型与隐式接口的完美折中】
 * 本例展示了 Go 语言如何通过“结构化类型（Structural Typing）”在静态安全与动态灵活之间找到平衡。
 * Go 允许在参数列表中内联定义匿名接口（interface { speak(); sit() }），且不需要具体类型（Dog/Robot）显式声明实现该接口（无 implements 关键字）。
 * 只要结构满足要求即视为匹配，这种“隐式契约”彻底打破了 Java 中实现类必须依赖接口定义的强耦合，是对 Java 繁琐的“名义类型系统”最现代化的降维打击。
 */
func perform(speaker interface { speak(); sit() }) {
  speaker.speak();
  speaker.sit();
}

func main() {
  perform(Dog{})
  perform(Robot{})
}
/* Output:
Arf!
Sitting
Click!
Clank!
*/
